src/wasi_unstable/mod.rs

```diff
@@ -49,9 +50,25 @@ pub type FileStat = __wasi_filestat_t;
 pub type CIoVec = __wasi_ciovec_t;
 pub type IoVec = __wasi_iovec_t;
 pub type Subscription = __wasi_subscription_t;
-pub type Event = __wasi_event_t;
 pub type Prestat = __wasi_prestat_t;
 
+// should have exactly the same layout as __wasi_event_t
+// TODO: add a more thorough test
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct Event {
+    pub userdata: __wasi_userdata_t,
+    pub res: Result<(), Error>,
```

Comment:
Is it valid to assume that `Result<(), Error>` has exactly the same layout as `__wasi_errno_t`? My interpretation of https://github.com/rust-lang/rust/issues/55615 suggests that this is not guaranteed to work.

Reply:
IIUC the main issue is that, strictly speaking, right now Rust does not guarante that niche optimization will be applied to `Result<(), NonZeroU16>`. But this is exactly why I've added the size assert. I will try to ask in the thread a question about our specific case.

Reply:
So it seems this code is safe in the presence of asserts which we have, but theoretically it may result in a compile-time error in future Rust versions (although such event is extremely unlikely). This is why we need an RFC which will set the niche optimization on which we rely here in stone.

I think we can proceed with this code, without waiting for the RFC.

Reply:
BTW `Event` contains field with `__wasi_event_u` type, so users will have to fall back to `raw` interfaces. Ideally we would like to define something like:
```rust
#[repr(u8, C)]
enum EventPayload {
    Clock = EVENTTYPE_CLOCK,
    Read {
        nbytes: FileSize,
        flags: EventRwFlags,
    } = EVENTTYPE_FD_READ,
    Write {
        nbytes: FileSize,
        flags: EventRwFlags,
    } = EVENTTYPE_FD_WRITE,
}

#[repr(u8, C)]
enum SubscriptionPayload {
    Clock {
        identifier: UserData,
        clock_id: ClockId,
        timeout: Timestamp,
        precision: Timestamp,
        flags: SubclockFlags,
    } = EVENTTYPE_CLOCK,
    Read { fd: Fd } = EVENTTYPE_FD_READ,
    Write { fd: Fd } = EVENTTYPE_FD_WRITE,
}
```

But unfortunately it would require a [nightly feature](https://github.com/rust-lang/rust/issues/60553) and I am not sure about layout guarantees.

So how about we will keep using `__wasi_event_t` for the time being?

Reply:
I'm a little confused; using a union requires `unsafe`, but that doesn't require falling back to the `raw` interfaces.

The assert may be enough for today's rustc, but is it enough for [miri](https://github.com/rust-lang/miri)? I'm nervous about including something which may be in violation of language rules, even if we sanity-check it with asserts, at such a fundamental piece of the overall stack.

Would it be ok to continue to expose `Event` and `Errno` as they are, even if we use `Result` in place of `Errno` for most return types, for now? We could provide functions which manipulate the fields of an `Event` with a safe API to eliminate that bit of unsafety.

And if you or anyone is interested in building a higher-level API around `poll_oneoff` which provides nicer enum types or other convenience features, that'd be cool, though I'd suggest creating a new module at that point, so that we preserve the minimally-safe API for users that want that.

Reply:
Yes, I've already reverted the `Result` change and `Event` is back to being a simple alias.